chains/evm/calls/transactor/monitored/monitored.go

```diff
@@ -0,0 +1,206 @@
+		value:        opts.Value,
+		gasLimit:     opts.GasLimit,
+		gasPrice:     gp,
+		data:         data,
+		submitTime:   time.Now(),
+		creationTime: time.Now(),
+	}
+	tx, err := t.txFabric(rawTx.nonce, rawTx.to, rawTx.value, rawTx.gasLimit, rawTx.gasPrice, rawTx.data)
+	if err != nil {
+		return &common.Hash{}, err
+	}
+
+	h, err := t.client.SignAndSendTransaction(context.TODO(), tx)
+	if err != nil {
+		return &common.Hash{}, err
+	}
+
+	t.txLock.Lock()
+	t.pendingTxns[h] = rawTx
+	t.txLock.Unlock()
+
+	err = t.client.UnsafeIncreaseNonce()
+	if err != nil {
+		return &common.Hash{}, err
+	}
+
+	return &h, nil
+}
+
+func (t *MonitoredTransactor) Monitor(
+	ctx context.Context,
+	resendInterval time.Duration,
+	txTimeout time.Duration,
+	tooNewTransaction time.Duration,
+) {
+	ticker := time.NewTicker(resendInterval)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			{
+				t.txLock.Lock()
+				pendingTxCopy := make(map[common.Hash]RawTx, len(t.pendingTxns))
+				for k, v := range t.pendingTxns {
+					pendingTxCopy[k] = v
+				}
+				t.txLock.Unlock()
+
+				for oldHash, tx := range pendingTxCopy {
+					receipt, err := t.client.TransactionReceipt(context.Background(), oldHash)
+					if err == nil {
+						if receipt.Status == types.ReceiptStatusSuccessful {
+							log.Info().Uint64("nonce", tx.nonce).Msgf("Executed transaction %s with nonce %d", oldHash, tx.nonce)
+						} else {
+							log.Error().Uint64("nonce", tx.nonce).Msgf("Transaction %s failed on chain", oldHash)
+						}
+
+						delete(t.pendingTxns, oldHash)
+						continue
+					}
+
+					if time.Since(tx.creationTime) > txTimeout {
+						log.Error().Uint64("nonce", tx.nonce).Msgf("Transaction %s has timed out", oldHash)
+						delete(t.pendingTxns, oldHash)
+						continue
+					}
+					if time.Since(tx.submitTime) < tooNewTransaction {
+						continue
+					}
+
+					hash, err := t.resendTransaction(&tx)
+					if err != nil {
+						log.Warn().Uint64("nonce", tx.nonce).Err(err).Msgf("Failed resending transaction %s", hash)
+						continue
+					}
+
+					delete(t.pendingTxns, oldHash)
+					t.pendingTxns[hash] = tx
+				}
+			}
+		}
+	}
+}
+
+func (t *MonitoredTransactor) resendTransaction(tx *RawTx) (common.Hash, error) {
+	tx.gasPrice = t.increaseGas(tx.gasPrice)
+	newTx, err := t.txFabric(tx.nonce, tx.to, tx.value, tx.gasLimit, tx.gasPrice, tx.data)
+	if err != nil {
+		return common.Hash{}, err
+	}
+
+	hash, err := t.client.SignAndSendTransaction(context.TODO(), newTx)
+	if err != nil {
+		return common.Hash{}, err
+	}
+
+	log.Debug().Uint64("nonce", tx.nonce).Msgf("Resent transaction with hash %s", hash)
+
+	return hash, nil
+}
+
+// increase gas bumps gas price by preset percentage.
+//
+// If gas was 10 and the increaseFactor is 15 the new gas price
+// would be 11 (it floors the value). In case the gas price didn't
+// change it increases it by 1.
+func (t *MonitoredTransactor) increaseGas(oldGp []*big.Int) []*big.Int {
```

Comment:
I think we should add unit tests for this function specifically to be able to determine changes in this part